openraft/src/core/mod.rs

```diff
@@ -449,8 +449,12 @@ impl<D: AppData, R: AppDataResponse, N: RaftNetwork<D>, S: RaftStorage<D, R>> Ra
                 // Transition to follower.
                 self.set_target_state(State::Follower);
             }
-        } else {
+        } else if self.effective_membership.membership.is_turn_to_learner() {
             self.set_target_state(State::Learner);
+            tracing::debug!("node {} turn to learner", self.id);
+        } else {
+            self.set_target_state(State::Shutdown);
+            tracing::debug!("node {} has been removed,shutdowning...", self.id);
```

Comment:
I do not quite get it:
Since `Membership` contains a specific list of members and a list of learners, we can just check the `node_id` against these two lists to decide what to do: to turn it to a learner or shut it down.